src/python/pants/backend/go/pkg.py

```diff
@@ -151,56 +222,58 @@ async def resolve_go_package(
     )
 
     metadata = json.loads(result.stdout)
-
-    # TODO: Raise an exception on errors. They are only emitted as warnings for now because the `go` tool is
-    # flagging missing first-party code as a dependency error. But we want dependency inference and won't know
-    # what the dependency actually is unless we first resolve the package with that dependency. So circular
-    # reasoning. We may need to hydrate the sources for all go_package targets that share a `go_module`.
-    if metadata.get("Incomplete"):
-        error_dict = metadata.get("Error", {})
-        if error_dict:
-            error_str = error_to_string(error_dict)
-            logger.warning(
-                f"Error while resolving Go package at address {request.address}: {error_str}"
-            )
-        # TODO: Check DepsErrors key as well.
-
-    # Raise an exception if any unsupported source file keys are present in the metadata.
-    for key in (
-        "CompiledGoFiles",
-        "CFiles",
-        "CXXFiles",
-        "MFiles",
-        "HFiles",
-        "FFiles",
-        "SFiles",
-        "SwigFiles",
-        "SwigCXXFiles",
-        "SysoFiles",
-    ):
-        files = metadata.get(key, [])
-        if files:
-            raise ValueError(
-                f"The go_package at address {request.address} contains the following unsupported source files "
-                f"that were detected under the key '{key}': {', '.join(files)}."
-            )
-
-    return ResolvedGoPackage(
-        address=request.address,
+    return ResolvedGoPackage.from_metadata(
+        metadata,
         import_path=import_path,
+        address=request.address,
         module_address=owning_go_module_result.module_address,
-        package_name=metadata["Name"],
-        imports=tuple(metadata.get("Imports", [])),
-        test_imports=tuple(metadata.get("TestImports", [])),
-        dependency_import_paths=tuple(metadata.get("Deps", [])),
-        go_files=tuple(metadata.get("GoFiles", [])),
-        cgo_files=tuple(metadata.get("CgoFiles", [])),
-        ignored_go_files=tuple(metadata.get("IgnoredGoFiles", [])),
-        ignored_other_files=tuple(metadata.get("IgnoredOtherFiles", [])),
-        test_go_files=tuple(metadata.get("TestGoFiles", [])),
-        xtest_go_files=tuple(metadata.get("XTestGoFiles", [])),
     )
 
 
+@dataclass(frozen=True)
+class ResolveExternalGoModuleToPackagesRequest:
+    path: str
+    version: str
+
+
+@dataclass(frozen=True)
+class ResolveExternalGoModuleToPackagesResult:
+    packages: FrozenOrderedSet[ResolvedGoPackage]
```

Comment:
Use `DeduplicatedCollection` from `pants.engine.collections`.

Reply:
I see it only used via subclassing and not as a type of a dataclass field. Is it okay to use as a type for a dataclass field?

Reply:
Going to add a TODO for now so I can land the PR, but will revisit during cleanup phase.

Reply:
Yeah it's only used for subclassing. So if you will need more than 1 field, it isn't appropriate.

Reply:
Okay will leave as FrozenOrderedSet for now as I may add additional fields to this dataclass.